### DISCUSSION TITLE
Cache incrémental

### DISCUSSION THREAD

**Discussion by mquandalle** [publicodes/publicodes#393] [Ideas]
Ce ticket s'inspire très largement des travaux sur la librairie [Incremental](https://github.com/janestreet/incremental), dont le modèle d'évaluation est comparable au notre. Plutôt que de ré-implémenter toutes nos logiques spécifiques (applicabilité, variables manquantes, inversions, recalculs, etc.) à partir des abstractions bas niveau d'Incremental, il propose d'amender notre logique existante pour y intégrer une invalidation de cache incrémentale.

L'implémentation de cette proposition devrait être raisonnablement simple, mais elle nécessite d'être coordonnée avec d'autres évolutions du moteur, en particulier betagouv/mon-entreprise#1191 et betagouv/publicodes#23.

## Granularité du cache

Aujourd'hui quand une même règle est référencée à plusieurs reprises dans l'arbre d'évaluation, elle est évaluée une première fois puis mise en cache dans une hashmap via `cache['nom de la règle']`. Elle est ensuite récupérée du cache lors des références ultérieures.

La première idée est de généraliser cette logique de cache en l'intégrant au niveau de chaque nœud et non plus des seules règles. Ainsi si un nœud variation `si: oui: alors: x sinon: y` a déjà été évalué à `x` on ne refait pas le calcul à chaque fois qu'on le rencontre. Cela est particulièrement critique pour la nouvelle implémentation des remplacements qui, parce qu'elle agit au niveau des références et non plus des règles, ajoute des nœuds `variations` en grande quantité dans l'arbre. Sans cache au niveau des nœuds le calcul subit une baisse de performance sensible.

À noter que les nœuds intermédiaires comme `variations` n'ont pas de nom et qu'on ne peut donc pas les référencer dans un cache distinct du type `cache['nom du noeud']`. On pourrait donner un nom à chaque nœud en lui ajoutant un identifiant `nodeId: i++` lors de la création puis aller chercher sa valeur dans `cache[44582]`, mais il est plus simple de ne plus distinguer le nœud évalué de son cache, et de lire et écrire l'attribut `nodeValue` directement dans l'arbre d'évaluation:

```js
evaluateNode(node) {
  if (node.isCached) {
    const { nodeValue } = evaluationFunctions[node.nodeKind](node)
    node.nodeValue = nodeValue
    node.isCached = true
  }
  return node
}
```

L'utilisation d'un cache au niveau des nœuds plutôt que des règles doit permettre de résoudre les problèmes de performance dans betagouv/mon-entreprise#1191. Mais pour que cela fonctionne, il faut que le mécanisme "remplace" référence une même instance du nœud "variations" partout où elle est utilisée plutôt que d'avoir des instances distincte à chaque utilisation. 

## Cache entre les situations

Actuellement on fait une remise à zéro du cache à chaque changement de situation. Cela nous oblige à refaire inutilement certains calculs dont nous ne pouvons détecter qu'ils ont déjà été effectués à l'identique dans la situation précédente. L'implémentation la plus simple du modèle d'Incremental consiste à ajouter un booléen `isDirty = true` sur chaque nœud déterminant s'il doit être évalué ou non; en reprenant l'exemple ci-dessus :

```js
evaluateNode(node) {
  if (node.isDirty) {
    const { nodeValue } = evaluationFunctions[node.nodeKind](node)
    node.nodeValue = nodeValue
    node.isDirty = false
  }
  return node
}
```

Lors d'un changement de situation on marque comme `isDirty = true` les nœuds terminaux qui ont été modifiés, ainsi que les nœuds qui en dépendent. Cela implique un parcours de l'arbre à partir des feuilles, où chaque étape doit être rapide car on ne fait qu'une comparaison de booléen et un éventuel changement de `false` à `true`. À noter que si on tombe sur une nœud avec l'attribut `isDirty == true`, on n'a pas besoin d'aller explorer ses parents et on peut "couper la branche". Cela assure notamment que si une partie de l'arbre n'est pas utilisée (exemple les règles des "artistes-auteurs" dans le simulateur salarié) on ne se baladera pas dans ce sous-arbre, les règles resteront `isDirty == true` tout au long des simulations successives.

```js
setSituation(newSituation) {
  const changedNodes = diff(this.situation, newSituation)
  changedNodes.forEach(node => setDirty(node))
}

setDirty(node) {
  if (!node.isDirty) {
    node.isDirty = true
    node.usedBy.forEach(node => setDirty(node))
  }
}
```

> Note: Il est possible d'ajouter une logique "d'observateur" afin de couper des branches supplémentaires mais cela ajoute de la complexité, et je pense que ça n'aurait de toute façon pas de gros impact vu que les valeurs que l'on chercher à calculer (les observateurs donc) sont relativement stables entre chaque simulation.

À noter aussi que la possibilité de réutiliser des calculs entre les simulations dépend intrinsèquement de la topologie des règles. La chose profitera ainsi largement aux simulateurs d'Ecolab où les branches principales sont indépendantes les unes des autres

```
impact total = impact nourriture + impact transport + impact maison
impact transport = impact voiture + impact avion
```

Ici si la modification de la valeur d'`impact voiture` entraîne un recalcul minimal (avion, nourriture et maison restant en cache).

Sur mon-entreprise, on a une règle centrale “assiette des cotisations sociales” impactée quasiment à chaque fois que l'on répond à une question, et dont la modification entraîne le recalcul de la majorité de l'arbre. Si je reçois par exemple un ticket restaurant, il faut recalculer l'assiette des cotisations, puis l'ensemble des cotisations, puis l'impôt sur le revenu. Néanmoins on profitera tout de même de cette logique à plusieurs endroits, par exemple le "plafond JEI" est une constante qui n'a pas besoin d'être recalculé à chaque changement de situation, de même le calcul des "notifications" est largement inchangé entre les simulations.

## Cache des inversions et recalculs

Vu leur fonctionnement spécifique les mécanismes inversions et recalculs nécessitent une attention particulière. Je réalise que mon ticket est déjà bien long, donc je fais vite sur ces deux mécanismes. Même si les détails devront être précisés pour l’implémentation, je ne vois aucune incompatibilité de principe entre leur sémantique et l'invalidation de cache incrémentale.

#### Recalculs

L'utilisation du cache par le mécanisme recalcul n'est pas compatible en l'état avec le modèle incrémental, car on modifie les valeurs existantes des nœuds pendant le recalcul, avant de restaurer le cache original. Il n'est en fait pas logique de réutiliser des nœuds existants si on sait que leur valeur sera différente, autrement dit les nœuds `assiette des cotisations` et `assiette des cotisations [dans le recalcul du plafond JEI]` doivent être 2 nœuds différents.

Je crois que la manière la plus simple et élégante de gérer ces nœuds "dans un recalcul" est d'avoir une instance distincte du moteur, instanciée au niveau du nœud `recalcul`. Lors de l'évaluation d'un recalcul, on fait un `setSituation` localement avec la situation amendée, puis un `evaluate` de l'objectif de recalcul. Pour des cas comme le "plafond JEI", le nœud n'aura pas été marqué comme `isDirty` et on pourra donc récupérera sa valeur dans le cache du sous-moteur sans calcul supplémentaire.

À noter que lors de l'exploration de la documentation, le mécanisme "recalcul" est actuellement une boîte noire. On voit seulement "cette valeur a été calculée en prenant telle situation en entrée". Avec la création d'une sous-instance du moteur au niveau du nœud recalcul, il devient possible "d'entrer" dans le sous-calcul et d'exposer dans la documentation toutes les valeurs intermédiaires utilisées.

#### Inversions

Même si une inversion est invalidée lors d'un changement de situation, elle peut bénéficier des calculs effectués précédemment pour accélérer sa recherche d'une solution. Le fait de garder les nœuds "vivants" plutôt qu'avoir une remise à zéro du cache à chaque changement de situation permettra donc d'implémenter une logique de ré-utilisation des calculs, cf. https://github.com/betagouv/mon-entreprise/pull/1231#issuecomment-732096305.

**Comment by denismerigoux** [comment]
Salut, tout ça va dans la bonne direction. Je vous donne un conseil : plutôt que de faire vous même un cache avec l'attribut `isDirty` qu'il va falloir gérer à la main, vous devriez utiliser des structures de données "persistantes" qui implémentent du partage de données sous le capot : https://github.com/immutable-js/immutable-js. 

Utiliser une structure de donnée persistante vous facilitera énormément la vie quand vous devrez faire du backtracking dans vos calculs, et ça vous permettra aussi de gérer les problèmes de scoping très naturellement. Utiliser des structures de données persistentes fait partie des bonnes pratiques de base quand on implémente un compilateur.

**Comment by mquandalle** [comment]
Oui j'ai bien conscience que les extraits de code de mon message utilisent un style impératif qui pose tout un tas de problèmes. Mais JavaScript ne proposant pas de structure immuable par défaut on est de toute façon obligé d'utiliser d'évaluer ce type de code, que ça soit directement ou par l'intermédiation d'une librairie comme `immutable-js`.

En général je pense qu'`immutable-js` est la bonne solution mais j'ai l'impression que pour certaines fonction cœur du moteur comme `evaluateNode` que l'on appelle des milliers de fois à chaque évaluation, il vaut mieux être le plus "proche du langage" possible, quitte à ré-implémenter nous-même une logique de "structural sharing". C'est un peu un bloc `unsafe` qu'on choisit d'optimiser pour les performances, par exemple quand je change le booléan `isDirty` je veux explicitement muter l'objet existant et non en créer une nouvelle instance. En revanche l'API exposée aux mécanismes ainsi que l'API publique doivent utilser des données immuables (ce que l'on fait aujourd'hui, toutes les fonctions retournent pas exemple des `{...baseNode, nodeValue }`), mais peut-être que la librairie pourrait avoir sa place à certains endroits.

**Comment by mquandalle** [comment]
Déjà cités ailleurs, mais l'inspiration vient de https://lord.io/blog/2020/spreadsheets/ et https://blog.janestreet.com/seven-implementations-of-incremental/

**Comment by denismerigoux** [comment]
Vous remarquerez que la librairie incremental de Janestreet est implémentée en OCaml :) Si vous voulez vraiment plus de performance vous feriez mieux de compiler à l'avance vos règles YAML en JS ou en Wasm directement pour éviter d'avoir un interpréteur. Parce qu'en réimplémentant vous-même le structural sharing vous risquez fort de vous planter beaucoup avant d'avoir un gain de performance très faible par rapport à utiliser `immutable-js`. En effet la gestion du contexte d’exécution et des variables c'est LE truc casse-gueule quand on écrit un interpréteur, avec à la clé des erreurs super subtiles à débugguer. Si vous voulez vous lancer dedans allez-y, je pense que vous y arriverez mais je vous préviens que ça va vous prendre pas mal de temps et que ça risque d'être un nid à bugs pour la suite de votre projet.

**Comment by johangirod** [comment]
See if this is needed once #724 is implemented
